### ISSUE TITLE
Dynamic Input and Output and data view from json data

### ISSUE THREAD

**Issue by oliver021** [dotnet/machinelearning#6022]
I have a problem that despite having some suggestions on how to solve it, the possible solutions are still a bit forced for my taste.
What happens is that `prediction engine` does not work to make predictions with an ML.Net model because both the input and the output are dynamically established in a schema stored in a data store, the problem is not to enter the model with data whose fields are dynamic, until I get there. The problem is when it comes to loading the model and using it to predict and make the work for which it was trained. I reiterate the schema of an input is strictly related to its model, but the problem is that the data to enter when predicting is data that is entered in a json. How can I from a json or a `Dictionary <string, string>` create an `IDataView` instance cleanly and without violating the API, to use `ITransform` directly to make predictions or uses of the model?

**Comment by luisquintanilla** [comment]
Hi @oliver021 

While not perfect, using the DataFrame API might be one of the simpler ways to do it. More details in this post.

https://www.luisquintanilla.me/posts/mlnet-predictions-no-schema.html

**Comment by michaelgsharp** [comment]
@oliver021 this is something that we are continuing to think about as its something we continue to hear. We are currently trying to find the right way to go about making this easier to use.

Were you able to test the solution suggested by luis?

**Comment by oliver021** [comment]
Thank you all for responsing. I made some interesting changes in my project, to work with dynamic inputs something like `Dynamic Prediction Engine`. Also I implement `IDataView` for anonymus object. 
For example: 
```cs
/// <summary>
    /// This implements a data view that has a schema, but no rows.
    /// </summary>
    public class DataViewCustom : IDataView
    {
        private readonly ISourceIterator iterator;

        //private readonly IHost _host;

        public bool CanShuffle => true;
        public DataViewSchema Schema { get; }

        /// <summary>
        /// Require schema and source
        /// </summary>
        /// <param name="schema"></param>
        /// <param name="iterator"></param>
        public DataViewCustom(DataViewSchema schema, ISourceIterator iterator)
        {
            if (iterator is null)
            {
                throw new ArgumentNullException(nameof(iterator));
            }

            //_host = env.Register(nameof(DataViewBase));
            Schema = schema;
            this.iterator = iterator;
        }

        /// <summary>
        /// Get number of the row at source
        /// </summary>
        /// <returns></returns>
        public long? GetRowCount() => iterator.Count();

        /// <summary>
        /// Get basic default cursor
        /// </summary>
        /// <param name="columnsNeeded"></param>
        /// <param name="rand"></param>
        /// <returns></returns>
        public DataViewRowCursor GetRowCursor(IEnumerable<DataViewSchema.Column> columnsNeeded, Random rand = null)
        {
            return new Cursor(iterator,Schema, columnsNeeded);
        }

        /// <summary>
        /// Just return one cursor for default
        /// </summary>
        /// <param name="columnsNeeded"></param>
        /// <param name="n"></param>
        /// <param name="rand"></param>
        /// <returns></returns>
        public DataViewRowCursor[] GetRowCursorSet(IEnumerable<DataViewSchema.Column> columnsNeeded, int n, Random rand = null)
        {
            return new[] { new Cursor(iterator, Schema, columnsNeeded) };
        }

        /// <summary>
        /// The correspond implementation
        /// </summary>
        private sealed class Cursor : DataViewRowCursor
        {
            /// <summary>
            /// Active columns for this cursor
            /// </summary>
            private readonly bool[] _active;

            /// <summary>
            /// the source iterator
            /// </summary>
            private readonly ISourceIterator _iterator;

            /// <summary>
            /// schema to model this cursor
            /// </summary>
            public override DataViewSchema Schema { get; }
            
            /// <summary>
            /// no paralell then default value is one
            /// </summary>
            public override long Batch => 1;
            
            /// <summary>
            /// Access to position
            /// </summary>
            public override long Position => _position;
            
            /// <summary>
            /// Declare variable to use to export through <see cref="Position"/>
            /// </summary>
            long _position;

            /// <summary>
            /// Helper to resolve fields
            /// </summary>
            /// <param name="columnsNeeded"></param>
            /// <param name="sourceSchema"></param>
            /// <param name="fieldUse"></param>
            static void ResolveActivesColumns(IEnumerable<DataViewSchema.Column> columnsNeeded, DataViewSchema sourceSchema, out bool[] fieldUse)
            {
                // set for default in false all columns
                fieldUse = new bool[sourceSchema.Count];
                foreach (var col in columnsNeeded)
                {
                    // security verification
                    if (col.Index >= fieldUse.Length)
                        throw new InvalidOperationException();

                    // set as active to use
                    fieldUse[col.Index] = true;
                }

            }

            /// <summary>
            /// Required iterator, requiremnts columns and schema
            /// </summary>
            /// <param name="iterator"></param>
            /// <param name="schema"></param>
            /// <param name="columnsNeeded"></param>
            public Cursor(ISourceIterator iterator, DataViewSchema schema, IEnumerable<DataViewSchema.Column> columnsNeeded)
            {
                if (iterator is null)
                {
                    throw new ArgumentNullException(nameof(iterator));
                }

                if (columnsNeeded is null)
                {
                    throw new ArgumentNullException(nameof(columnsNeeded));
                }

                _iterator = iterator;
                Schema = schema;

                // determine active columns by need columns passed
                ResolveActivesColumns(columnsNeeded, schema, out _active);
            }

            /// <summary>
            /// Std implementation of GetIdGetter
            /// </summary>
            /// <returns></returns>
            public override ValueGetter<DataViewRowId> GetIdGetter()
            {
                // return getter for id by position
                return (ref DataViewRowId val) => {
                    // id position
                    val = new DataViewRowId((ulong)Position, 0);
                };
            }

            /// <summary>
            /// Returns whether the given column is active in this row.
            /// </summary>
            public override bool IsColumnActive(DataViewSchema.Column column) =>
                column.Index < _active.Length && _active[column.Index];

            /// <summary>
            /// Returns a value getter delegate to fetch the value of column with the given columnIndex, from the row.
            /// This throws if the column is not active in this row, or if the type
            /// <typeparamref name="TValue"/> differs from this column's type.
            /// </summary>
            /// <typeparam name="TValue"> is the column's content type.</typeparam>
            /// <param name="column"> is the output column whose getter should be returned.</param>
            public override ValueGetter<TValue> GetGetter<TValue>(DataViewSchema.Column column)
            {
                return (ref TValue value) => {
                    value = (dynamic)_iterator.GetValue(column.Name, column.Type.RawType);
                };
            }

            /// <summary>
            /// Next element
            /// </summary>
            /// <returns></returns>
            public override bool MoveNext()
            {
                // increment position
                _position += 1;

                // pass to the next row
                return _iterator.Next();
            }
        }
    }

```

The `ISourceIterator` can be any model or data structure with one or more rows. For example, one of my implementations is json `IDataView` that recive direct json text source.  So please I love any feedback on the code above.

**Comment by oliver021** [comment]
I did some tests and it worked perfectly, because the problem is that a json document comes to me, where a field tells me which model to use and if it is available, apply it to the rest of the data that comes in the json. Since models are added on the fly to an already deployed system, I can't have classes as schemas for those models. The key is that with `IJsonDataView` I was able to fix this problem.
